ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java

```diff
@@ -0,0 +1,612 @@
+   * @throws Exception
+   */
+  void checkOperationChains(List<OperationChain> opChains)
+      throws Exception {
+    Map<PartitionId, Integer> partitionCount = new HashMap<>();
+    double blobsPut = 0;
+    for (OperationChain opChain : opChains) {
+      opChain.latch.await(AWAIT_TIMEOUT, TimeUnit.SECONDS);
+      synchronized (opChain.testFutures) {
+        for (TestFuture testFuture : opChain.testFutures) {
+          testFuture.check();
+        }
+      }
+      if (opChain.blobId != null) {
+        blobsPut++;
+        PartitionId partitionId = new BlobId(opChain.blobId, clusterMap).getPartition();
+        int count = partitionCount.containsKey(partitionId) ? partitionCount.get(partitionId) : 0;
+        partitionCount.put(partitionId, count + 1);
+      }
+    }
+    double numPartitions = clusterMap.getWritablePartitionIds().size();
+    if (opChains.size() > numPartitions) {
+      double blobBalanceThreshold = 3.0 * Math.ceil(blobsPut / numPartitions);
+      for (Map.Entry<PartitionId, Integer> entry : partitionCount.entrySet()) {
+        Assert.assertTrue("Number of blobs is " + entry.getValue() + " on partition: " + entry.getKey()
+                + ", which is greater than the threshold of " + blobBalanceThreshold,
+            entry.getValue() <= blobBalanceThreshold);
+      }
+    }
+  }
+
+  /**
+   * Create an {@link OperationChain} from a queue of {@link OperationType}s.  Start the operation chain asynchronously.
+   * @param blobSize the size of the blob generated for put operations.
+   * @param chainId a numeric identifying the operation chain.
+   * @param operations the queue of operations to perform in the chain
+   * @return an {@link OperationChain} object describing the started chain.
+   */
+  OperationChain startOperationChain(int blobSize, int chainId, Queue<OperationType> operations) {
+    byte[] userMetadata = new byte[1000];
+    byte[] data = new byte[blobSize];
+    new Random().nextBytes(userMetadata);
+    new Random().nextBytes(data);
+    BlobProperties properties = new BlobProperties(blobSize, "serviceid1");
+    OperationChain opChain = new OperationChain(chainId, properties, userMetadata, data, operations);
+    continueChain(opChain);
+    return opChain;
+  }
+
+  /**
+   * Generate the properties needed by the router and coordinator.  NOTE: Properties for SSL interaction need
+   * to be added manually.
+   * @param routerDatacenter the datacenter name where the router will be running.
+   * @return a {@link Properties} object with the properties needed to instantiate the router/coordinator.
+   */
+  static Properties getRouterProperties(String routerDatacenter) {
+    Properties properties = new Properties();
+    properties.setProperty("router.hostname", "localhost");
+    properties.setProperty("router.datacenter.name", routerDatacenter);
+    properties.setProperty("router.connection.checkout.timeout.ms", "5000");
+    properties.setProperty("router.request.timeout.ms", "20000");
```

Comment:
I had set a large timeout because multi-chunk operations were taking a long time.  I can make this smaller because I made the large blobs smaller to speed up the test.